platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,270 @@
+
+### Requirements
+
+- Introduce an interface representing state lifecycle management and its implementation.
+- Update the current State API to align with the State Lifecycle API.
+- Implement these APIs and migrate existing code to use them.
+
+## Design & Architecture
+
+### Summary
+
+The following design describes a solution for managing the lifecycle of the state object. It covers the following aspects:
+
+- State initialization (loading a snapshot or creating a new state)
+- Maintaining references to the mutable state and the latest immutable state
+- Restricting mutability to a single state object
+- Loading snapshots from disk
+
+## Current State of the Code
+
+### Reservation Count Mechanism
+
+This section provides important context about how state objects are managed in memory.
+
+The reservation count mechanism ensures that an object is not garbage collected until it is no longer needed.
+This mechanism is implemented using a combination of the `Reservable` and `Releasable` interfaces.
+A state object, like any other Merkle tree node, can be reserved and released. The number of reservations and releases is tracked via a reference count.
+
+An important concept to understand with this mechanism, and the classes implementing these interfaces,
+is the distinction between **implicit reservations** and **explicit reservations**.
+
+#### Implicit Reservations
+
+When an object is initially constructed, it is considered to have an implicit reservation.
+Although its reservation count is 0, it will not be garbage collected because the caller of the constructor
+is presumed to still need the object. The reservation count for an object with an implicit reservation is 0.
+If `Releasable#release()` is called on such an object, it will be destroyed.
+
+#### Explicit Reservations
+
+When an object with an implicit reservation has `Reservable#reserve()` called on it for the first time, the reservation
+becomes explicit, and its reservation count increases to 1. From this point forward, the reservation remains
+explicit until the object is destroyed. An object with an explicit reservation cannot return to having an implicit reservation.
+
+If, at any time after obtaining an explicit reservation, the object has `Releasable#release()` called enough times
+to reduce its reservation count to 0, the object will be destroyed, and its reservation count will be set to -1.
+
+---
+
+We rely on this mechanism to manage the number of states kept in memory.
+
+**Important**: It is the responsibility of the client code to ensure that the state is released once it is no longer needed.
+
+Once the reservation count reaches 0, the state becomes eligible for garbage collection.
+
+---
+
+### Java Classes
+
+`com.swirlds.state.State` provides the following lifecycle-related methods:
+
+- `State copy()`: Creates a mutable copy of the state. The source state becomes immutable.
+- `void computeHash()`: Computes the hash of the immutable state.
```

Comment:
Compute the state hash. Can only be called on immutable states

Reply:
For information reason why can it only be called on immutable state?
Is there consideration for being able to compute the hash in the mutable state?
Say I start with State A, do a `copy()`, apply state changes to the mutable state and want to capture the hash of the state at that point.
Is this feasible and supported?